### PR TITLE
Add antipattern note to jsonb_agg

### PR DESCRIPTION
**Pull request by JLDLaughlin**

### CODE REVIEW

**Comment by frankmcsherry** [comment]
If you don't mind, I'd be up for saying a bit more (rather than just trashing the method). What do you think about

NOTE: This function conceals fine-grained input updates, by translating individual input updates to entire record changes in the output. Any downstream computation on the results (for example, filtering, transformation, projection) may be more efficient if lifted before this method.

**Comment by rjnn** [comment]
Here's my take on the points to emphasize here, although I'm not sure as to the best way to *communicate* these points:

1) materializing `jsonb_agg(foo)` is an anti-pattern in that any changes from `foo` to `foo'` cause `jsonb_agg(foo)` to get thrown away completely and `jsonb_agg(foo')` to be created anew. This causes a lot of thrashing, in particular if you do anything _downstream_ of the `jsonb_agg`.

2) Nevertheless, some users will want to receive `jsonb_agg(foo)`. Hence we support it!

3) The cleanest recommendation is thus to:
Materialize a view all the way to `foo`: (e.g. `CREATE MATERIALIZED VIEW raw AS foo...`) so that as much work as possible is done eagerly, incrementally, and efficiently .

Then, create a non-materialized view wrapping the whole thing up: `CREATE VIEW agged as jsonb_agg(foo)`

Let the user query `agged`, so that the user gets it in the format that they desire (the jsonb aggregation), and the bit of code that aggs everything up is run on-demand, rather than flapping objects unnecessarily.

Is this fair? It might be wrong, and itself a different anti-pattern, so this is as much a question as it is a suggestion.

**Comment by JLDLaughlin** [comment]
@frankmcsherry @rjnn what do you think of the update?

**Comment by frankmcsherry** [comment]
Seems good. I think all we can hope for here is getting performance minded folks to ask questions, as it is quite a lot of learnings to stash in such a short message.

**Comment by rjnn** [comment]
Lgtm!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by benesch** [comment]
@sploiselle may want to weigh in on the formatting here. The tables aren't particularly well suited to blurbs this long. But not sure there's anywhere else to put the blurb at the moment. For easy viewing: https://deploy-preview-2563--materializeinc.netlify.com/docs/sql/functions/#aggregate-func

**Comment by sploiselle** [comment]
My strong preference for this would be to break out `json_agg` into its own doc. I'm glad to create the doc that contains the information we've discussed in this issue.

**Comment by JLDLaughlin** [comment]
@sploiselle I took a stab at it! could you let me know if that's what you were thinking? and, also if you have any thoughts on the text itself!

File: www/data/sql_funcs.json [hunk 1]
```diff
@@ -64,7 +64,8 @@
             },
             {
                 "signature": "jsonb_agg(expression) -> jsonb",
-                "description": "Aggregate values (including nulls) as a jsonb array"
+                "description": "Aggregate values (including nulls) as a jsonb array.",
+                "url": "/docs/sql/functions/jsonb_agg"
```
Comment (sploiselle) [inline www/data/sql_funcs.json:68]: this should be `"url": "jsonb_agg"`

**Pull request merged**